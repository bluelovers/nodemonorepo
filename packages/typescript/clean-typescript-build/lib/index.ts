import {partition} from 'ramda'
import {unlink, lstat, existsSync} from 'fs-extra'
import {traverse} from 'fs-tree-utils'

import {
  Options,
  TargetList,
  Clean
} from './types'

import {
  DEFAULT_DEEP_FUNC,
  DEFAULT_SOURCE_DETECTOR,
  DEFAULT_TARGET_SPECIFIER
} from './contants'

export * from './types'
export * from './contants'

/**
 * Delete all compiled products corresponding to source files inside `root`
 * @param root Directory that contains all source files
 * @param options Options
 * @param options.deep A function that tells when to go deeper
 * @param options.isSource A function that tells which file is a source file
 * @param options.listTargets A function that lists compiled products corresponding to a source file
 * @returns An object
 *   * Property `targets`: An aggregation of `options.listTargets`'s return values
 *   * Property `reports`: An array of reports which tell whether or not deletion succeed
 *   * Property `success`: An array of paths to files which are successfully deleted
 *   * Property `failure`: An array of paths to files which are failed to be deleted
 */
export async function clean (root: string, options?: Options): Promise<Clean.Result> {
  const del = (file: string) => unlink(file).then(onUnlinkSuccess, onUnlinkFailure)
  const r2f = (reports: Clean.Result.ReportList) => reports.map(x => x.file)
  const onUnlinkSuccess = () => ({success: true as true})
  const onUnlinkFailure = (error: any) => ({success: false as false, error})
  const targets = await listAllTargets(root, options)
  const reports = await Promise.all(targets.map(async file => ({file, deletion: await del(file)})))
  const [success, failure] = partition(x => x.deletion.success, reports)
  return {targets, reports, success: r2f(success), failure: r2f(failure)}
}

/**
 * List compiled products corresponding to source files inside `root`
 * @param root Directory that contains all source files
 * @param options Options
 * @param options.deep A function that tells when to go deeper
 * @param options.isSource A function that tells which file is a source file
 * @param options.listTargets A function that lists compiled products corresponding to a source file
 * @returns A promise that resolves a list of target file paths
 */
export async function listAllTargets (
  root: string,
  options: Options = {}
): Promise<TargetList> {
  const {
    deep = DEFAULT_DEEP_FUNC,
    isSource = DEFAULT_SOURCE_DETECTOR,
    listTargets = DEFAULT_TARGET_SPECIFIER
  } = options

  const existed = (await traverse(root, {deep, stat: lstat}))
    .filter(isSource)
    .map(listTargets)
    .reduce((prev, current) => [...prev, ...current], [])
    .filter(item => existsSync(item))

  const files = (await Promise.all(
    existed.map(async item => ({
      item,
      stats: await lstat(item)
    }))
  ))
    .filter(x => x.stats.isFile())
    .map(x => x.item)

  return files
}
