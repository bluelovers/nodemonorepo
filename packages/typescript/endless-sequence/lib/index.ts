import EndlessIterable from './iterable'
import * as applications from './applications'
export {EndlessIterable, applications}
