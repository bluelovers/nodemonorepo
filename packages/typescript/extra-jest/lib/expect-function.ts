export type TargetFunction<Return> = () => Return

interface Success<Result> {
  readonly thrown: false
  readonly result: Result
}

interface Failure<Error> {
  readonly thrown: true
  readonly error: Error
}

type ExecutionReport<Result, Error> = Success<Result> | Failure<Error>

export class FunctionMatchers<Return = any> {
  constructor (private fn: TargetFunction<Return>) {}

  private execute (): ExecutionReport<Return, any> {
    try {
      return { thrown: false, result: this.fn() }
    } catch (error) {
      return { thrown: true, error }
    }
  }

  private match<E, R> (
    handleFailure: (error: any) => E,
    handleSuccess: (value: Return) => R
  ) {
    const report = this.execute()
    return report.thrown ? handleFailure(report.error) : handleSuccess(report.result)
  }

  public get throws () {
    return this.match(x => expect(x), value => {
      throw new Error(
        `Expecting the matched function to throw an error, but it returns a value: ${value}`
      )
    })
  }

  public get returns () {
    return this.match(error => {
      throw new Error(
        `Expecting the matched function to return a value, but it throws an error: ${error}`
      )
    }, x => expect(x))
  }
}

export function expectFunction<Return = any> (fn: TargetFunction<Return>) {
  return new FunctionMatchers(fn)
}

export default expectFunction
