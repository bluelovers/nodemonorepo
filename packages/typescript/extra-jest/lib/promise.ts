export function snapRejected<X> (promise: Promise<X>) {
  return promise.then(
    () => {
      throw new Error(`Expecting promise to reject but it didn't`)
    },
    reason => {
      expect(reason).toMatchSnapshot()
    }
  )
}
