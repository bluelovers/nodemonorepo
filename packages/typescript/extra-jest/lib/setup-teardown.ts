import * as process from 'process'
import * as fsx from 'fs-extra'
import * as fsTreeUtils from 'fs-tree-utils'
import tempPath from 'unique-temp-path'
import {TreeObject} from 'fs-tree-utils'

export type PromiseFunc<X, Y> = (x: X) => Promise<Y>
export type SyncFunc<X, Y> = (x: X) => Y
export type SetupFunc<Y> = () => Promise<Y>
export type TeardownFunc<X> = (x: X) => Promise<void>
export type CalledFunc<X, Y> = PromiseFunc<X, Y>
export type SyncCalledFunc<X, Y> = SyncFunc<X, Y>
export type Tester = () => Promise<void>
export type AsyncTesterFactory<SM, MT> = (fn: CalledFunc<SM, MT>) => Tester
export type SyncTesterFactory<SM, MT> = (fn: SyncCalledFunc<SM, MT>) => Tester

export type Config<SY, TX> = {
  setup: SetupFunc<SY>, teardown: TeardownFunc<TX>
}

export type TesterFactory<SM, MT> = AsyncTesterFactory<SM, MT> & {
  forAsync: AsyncTesterFactory<SM, MT>,
  forSync: SyncTesterFactory<SM, MT>
}

export namespace base {
  export function createFactory<SM, MT> (
    {
      setup,
      teardown
    }: Config<SM, MT>
  ): TesterFactory<SM, MT> {
    const forAsync = (fn: CalledFunc<SM, MT>): Tester =>
      async (): Promise<void> => {
        const sm = await setup()
        const mt = await fn(sm)
        await teardown(mt)
      }

    const forSync = (fn: SyncCalledFunc<SM, MT>): Tester =>
      forAsync(async (sm: SM) => fn(sm))

    return Object.assign(
      forAsync,
      {
        forAsync,
        forSync
      }
    )
  }
}

export const createFactory = base.createFactory

export namespace virtualEnvironment {
  export interface Info {
    readonly tree: TreeObject
    readonly container: string
    readonly previousWorkingDirectory: string
  }

  export function createFactory (tree: TreeObject, container = tempPath()) {
    const previousWorkingDirectory = process.cwd()
    const info: Info = {tree, container, previousWorkingDirectory}

    async function populate () {
      await depopulate()
      await fsTreeUtils.create(tree, container)
      return info
    }

    async function depopulate () {
      await fsx.remove(container)
      return info
    }

    const setup: SetupFunc<Info> = async () => {
      await populate()
      process.chdir(container)
      return info
    }

    const teardown: TeardownFunc<void> = async () => {
      process.chdir(previousWorkingDirectory)
      await depopulate()
    }

    const apply = base.createFactory({setup, teardown})

    return {info, apply}
  }
}

export default createFactory
