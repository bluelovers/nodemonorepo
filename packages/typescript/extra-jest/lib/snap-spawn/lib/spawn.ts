import { SpawnSyncOptions, SpawnSyncReturns, spawnSync } from 'child_process'
import ramda from 'ramda'

export type IOData = Buffer | string
export type OptionalIOData = IOData | null | undefined

const fmtNull = <X>(x: X) =>
  x == null ? null : x

const fmtStdIO = (buf: OptionalIOData): string | null => {
  if (buf == null) return null
  const str = String(buf)
  return str.trim() ? `\n${str}\n` : '((EMPTY))'
}

export type Argv = string[]

export interface SpawnReturns {
  readonly status: number | null
  readonly signal: string | null
  readonly error: Error | null
  readonly stdout: string | null
  readonly stderr: string | null
}

export type SpawnFunc = (argv: Argv, options: SpawnSyncOptions) => SpawnSyncReturns<IOData>

export const createSpawnFunc = (command: string): SpawnFunc => ramda.partial(spawnSync, [command])

export function spawn (
  fn: SpawnFunc,
  argv: Argv = [],
  options: SpawnSyncOptions = {}
): SpawnReturns {
  const { status, signal, error, stdout, stderr } = fn(
    argv,
    options
  )

  return {
    status: fmtNull(status),
    signal: fmtNull(signal),
    error: fmtNull(error),
    stdout: fmtStdIO(stdout),
    stderr: fmtStdIO(stderr)
  }
}

export namespace spawn {
  export const withCommand = (
    command: string,
    argv?: Argv,
    options?: SpawnSyncOptions
  ) => spawn(createSpawnFunc(command), argv, options)
}

export default spawn
