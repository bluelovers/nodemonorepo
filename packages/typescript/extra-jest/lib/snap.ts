import * as yaml from 'js-yaml'

export type DumpOptions = yaml.DumpOptions
export type DumpFunction<X> = (x: X, o?: DumpOptions) => string
export type TestFunction<X> = (x: X, o?: DumpOptions) => () => void

export function mkfn<X> (fn: DumpFunction<X>, options: DumpOptions = {}): TestFunction<X> {
  return (x: X, extraOptions: DumpOptions = {}) => () => {
    expect('\n' + fn(x, { ...options, ...extraOptions }) + '\n').toMatchSnapshot()
  }
}

const defaultOptions: DumpOptions = {
  skipInvalid: true,
  sortKeys: true,
  noRefs: false,
  noCompatMode: true
}

const unsafeDump: DumpFunction<any> = (x, o) => yaml.dump(x, o)
const safeDump: DumpFunction<any> = (x, o) => yaml.safeDump(x, o)

export const unsafe = mkfn(unsafeDump, defaultOptions)
export const safe = mkfn(safeDump, defaultOptions)
export const pureUnsafe = mkfn(unsafeDump)
export const pureSafe = mkfn(safeDump)
export const noRefs = mkfn(unsafeDump, { noRefs: true })

export default unsafe
