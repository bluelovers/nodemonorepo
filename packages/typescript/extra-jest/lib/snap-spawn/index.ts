import {SpawnSyncOptions} from 'child_process'
import * as xsnap from '../snap'
import spawn, {SpawnFunc, createSpawnFunc} from './lib/spawn'

export * from './lib/spawn'

export function snap (
  fn: SpawnFunc,
  argv?: string[],
  options?: SpawnSyncOptions,
  snap = xsnap.safe
) {
  return snap(spawn(fn, argv, options))
}

export namespace snap {
  export const withCommand = (
    command: string,
    argv?: string[],
    options?: SpawnSyncOptions,
    fsnap = xsnap.safe
  ) => snap(createSpawnFunc(command), argv, options, fsnap)
}

export default snap
