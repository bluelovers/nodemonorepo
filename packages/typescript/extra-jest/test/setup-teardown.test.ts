import * as subject from '../lib/setup-teardown'

it('subject matches snapshot', () => {
  expect(subject).toMatchSnapshot()
})

describe('trackedData', () => {
  const initialTrackedData = 'Initial'
  const newTrackedData = 'New'
  let trackedData = initialTrackedData

  const createHistoryRecorder = (history: any[]) =>
    (step: 'setup' | 'teardown' | 'main') =>
      history.push({ trackedData, step })

  it('changes when invoking setup-teardown asynchronously', async () => {
    let history: any[] = []
    const record = createHistoryRecorder(history)

    const setup = async () => {
      record('setup')
      trackedData = newTrackedData
    }

    const teardown = async () => {
      record('teardown')
      trackedData = initialTrackedData
    }

    const main = async () => {
      record('main')
    }

    await subject.base.createFactory<void, void>({ setup, teardown }).forAsync(main)()
    expect(trackedData).toBe(initialTrackedData)
    expect({ history }).toMatchSnapshot()
  })

  it('changes when invoking setup-teardown synchronously', async () => {
    let history: any[] = []
    const record = createHistoryRecorder(history)

    const setup = async () => {
      record('setup')
      trackedData = newTrackedData
    }

    const teardown = async () => {
      record('teardown')
      trackedData = initialTrackedData
    }

    const main = () => {
      record('main')
    }

    await subject.base.createFactory<void, void>({ setup, teardown }).forSync(main)()
    expect(trackedData).toBe(initialTrackedData)
    expect({ history }).toMatchSnapshot()
  })

  it('stays unchanged when not invoking setup-teardown', () => {
    expect(trackedData).toBe(initialTrackedData)
  })
})
