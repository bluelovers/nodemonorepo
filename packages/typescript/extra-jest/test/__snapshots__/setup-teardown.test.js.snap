// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subject matches snapshot 1`] = `
Object {
  "base": Object {
    "createFactory": [Function],
  },
  "createFactory": [Function],
  "default": [Function],
  "virtualEnvironment": Object {
    "createFactory": [Function],
  },
}
`;

exports[`trackedData changes when invoking setup-teardown asynchronously 1`] = `
Object {
  "history": Array [
    Object {
      "step": "setup",
      "trackedData": "Initial",
    },
    Object {
      "step": "main",
      "trackedData": "New",
    },
    Object {
      "step": "teardown",
      "trackedData": "New",
    },
  ],
}
`;

exports[`trackedData changes when invoking setup-teardown synchronously 1`] = `
Object {
  "history": Array [
    Object {
      "step": "setup",
      "trackedData": "Initial",
    },
    Object {
      "step": "main",
      "trackedData": "New",
    },
    Object {
      "step": "teardown",
      "trackedData": "New",
    },
  ],
}
`;
